src/components/network/network-loader.js

```diff
@@ -19,117 +19,264 @@ genotet.utils.inherit(genotet.NetworkLoader, genotet.ViewLoader);
- * This function sets the gene regex to a verbose concatenation of the gene
- * ids. This is only called in removeGenes_().
+ * Deletes genes from graph and related edges.
+ * @param {!Array<string>} genes Names of the genes.
  * @private
  */
-genotet.NetworkLoader.prototype.updateGeneRegex_ = function() {
-  var regex = '';
-  this.data.nodes.forEach(function(node, index) {
-    regex += node.id + (index == this.data.nodes.length - 1 ? '' : '|');
+genotet.NetworkLoader.prototype.deleteGenes_ = function(genes) {
+  var geneMap = genotet.utils.keySet(genes);
+  // delete the nodes
+  for (var i = 0; i < this.data.network.nodes.length; i++) {
+    if (this.data.network.nodes[i].id in geneMap) {
+      this.data.network.nodes.splice(i, 1);
+    }
+  }
+  // delete the edges
+  for (var i = this.data.network.edges.length - 1; i >= 0; i--) {
+    if (this.data.network.edges[i].source in geneMap ||
+      this.data.network.edges[i].target in geneMap) {
+      this.data.network.edges.splice(i, 1);
+    }
+  }
+};
+
+/**
+ * Adds edges to the network.
+ * @param {!Array<!genotet.NetworkEdge>} edges Edges to be add into network.
+ */
+genotet.NetworkLoader.prototype.addEdges = function(edges) {
+  edges.forEach(function(edge) {
+    var source = edge.source;
+    var target = edge.target;
+    var weight = edge.weight;
+    var sourceExists = false, targetExists = false;
+    this.data.network.nodes.forEach(function(node) {
+      if (source == node.id) {
+        sourceExists = true;
+      }
+      if (target == node.id) {
+        targetExists = true;
+      }
+    });
+    if (!sourceExists) {
+      this.data.network.nodes.push({
+        id: source,
+        label: source,
+        isTF: true
+      });
+    }
+    if (!targetExists) {
+      var isTF = false;
+      this.data.networkInfo.nodes.forEach(function(node) {
+        if (target == node.id) {
+          isTF = node.isTF;
+        }
+      });
+      this.data.network.nodes.push({
+        id: target,
+        label: target,
+        isTF: isTF
+      });
+    }
+    this.data.network.edges.push({
+      id: source + ',' + target,
+      source: source,
+      target: target,
+      weight: weight
+    });
   }, this);
-  this.data.geneRegex = regex;
 };
 
 /**
- * Fetches the incident edges of a given node.
- * @param {!Object} node Node of which the incident edges are queried.
+ * Deletes edges from the network.
+ * @param {!Array<!genotet.NetworkEdge>} edges Edges to be deleted.
  */
-genotet.NetworkLoader.prototype.incidentEdges = function(node) {
-  var params = {
-    type: 'incident-edges',
-    fileName: this.data.fileName,
-    gene: node.id
-  };
-  this.get(genotet.data.serverURL, params, function(data) {
-    this.data.incidentEdges = data;
-    this.signal('incidentEdges');
-  }.bind(this), 'cannot get incident edges');
+genotet.NetworkLoader.prototype.deleteEdges = function(edges) {
+  edges.forEach(function(edge) {
```

Comment:
Can you create an object indexing the edge id's and then complete the deletion in O(nq) time (q is the time for accessing the object attribute, n is the number of edges in this.data.network.edges)?